packages/babel-helper-member-expression-to-functions/src/index.js

```diff
@@ -37,9 +86,153 @@ const handle = {
+      // Transforming optional chaining requires we replace ancestors.
+      if (isInDetachedTree(member)) return;
+
+      // We're looking for the end of _this_ optional chain, which is actually
+      // the "rightmost" property access of the chain. This is because
+      // everything up to that property access is "optional".
+      //
+      // Let's take the case of `FOO?.BAR.baz?.qux`, with `FOO?.BAR` being our
+      // member. The "end" to most users would be `qux` property access.
+      // Everything up to it could be skipped if it `FOO` were nullish. But
+      // actually, we can consider the `baz` access to be the end. So we're
+      // looking for the nearest optional chain that is `optional: true`.
+      const endPath = member.find(({ node, parent, parentPath }) => {
+        if (parentPath.isOptionalMemberExpression()) {
+          // We need to check `parent.object` since we could be inside the
+          // computed expression of a `bad?.[FOO?.BAR]`. In this case, the
+          // endPath is the `FOO?.BAR` member itself.
+          return parent.optional || parent.object !== node;
+        }
+        if (parentPath.isOptionalCallExpression()) {
+          // Checking `parent.callee` since we could be in the arguments, eg
+          // `bad?.(FOO?.BAR)`.
+          return parent.optional || parent.callee !== node;
+        }
+        return true;
+      });
+
+      const rootParentPath = endPath.parentPath;
+      if (
+        rootParentPath.isUpdateExpression({ argument: node }) ||
+        rootParentPath.isAssignmentExpression({ left: node })
+      ) {
+        throw member.buildCodeFrameError(`can't handle assignment`);
+      }
+      if (rootParentPath.isUnaryExpression({ operator: "delete" })) {
+        throw member.buildCodeFrameError(`can't handle delete`);
+      }
+
+      if (node.optional) {
+        throw member.buildCodeFrameError(
+          `can't handle '?.' directly before ${node.property.type}`,
+        );
+      }
+
+      // Now, we're looking for the start of this optional chain, which is
+      // optional to the left of this member.
+      //
+      // Let's take the case of `foo?.bar?.baz.QUX?.BAM`, with `QUX?.BAM` being
+      // our member. The "start" to most users would be `foo` object access.
+      // But actually, we can consider the `bar` access to be the start. So
+      // we're looking for the nearest optional chain that is `optional: true`,
+      // which is guaranteed to be somewhere in the object/callee tree.
+      let startingOptional = member;
+      for (;;) {
+        if (startingOptional.isOptionalMemberExpression()) {
+          if (startingOptional.node.optional) break;
+          startingOptional = startingOptional.get("object");
+          continue;
+        } else if (startingOptional.isOptionalCallExpression()) {
+          if (startingOptional.node.optional) break;
+          startingOptional = startingOptional.get("callee");
+          continue;
+        }
+      }
```

Comment:
Can we add a `throw new Error("Internal error")` at the end of this loop's body? It's better to throw than to accidentally have an infinite loop.